keylime-agent/src/main.rs

```diff
@@ -935,17 +737,95 @@ async fn main() -> Result<()> {
 
     let server_handle = server.handle();
     let server_task = rt::spawn(server).map_err(Error::from);
-    let worker_task = rt::spawn(worker(
-        symm_key,
-        symm_key_cvar,
-        payload,
+
+    // Only run payload scripts if mTLS is enabled or 'enable_insecure_payload' option is set
+    let run_payload = config.agent.enable_agent_mtls
+        || config.agent.enable_insecure_payload;
+
+    let payload_task = rt::spawn(payloads::worker(
         config.clone(),
         PathBuf::from(&mount),
+        payload_rx,
+        revocation_tx.clone(),
+        #[cfg(feature = "with-zmq")]
+        zmq_tx.clone(),
+    ))
+    .map_err(Error::from);
+
+    let key_task = rt::spawn(keys_handler::worker(
+        run_payload,
+        config.agent.uuid.clone(),
+        keys_rx,
+        payload_tx.clone(),
     ))
     .map_err(Error::from);
 
-    let result = try_join!(server_task, worker_task);
-    server_handle.stop(true).await;
+    // If with-zmq feature is enabled, run the service listening for ZeroMQ messages
+    #[cfg(feature = "with-zmq")]
+    let zmq_task = if config.agent.enable_revocation_notifications {
+        let ip = if let Some(i) = config.agent.revocation_notification_ip {
+            i
+        } else {
+            error!("No IP set in 'revocation_notification_ip' option");
+            return Err(Error::Configuration(
+                "No IP set in 'revocation_notification_ip' option"
+                    .to_string(),
+            ));
+        };
+
+        let port = if let Some(p) = config.agent.revocation_notification_port
+        {
+            p
+        } else {
+            error!("No port set in 'revocation_notification_port' option");
+            return Err(Error::Configuration(
+                "No port set in 'revocation_notification_port' option"
+                    .to_string(),
+            ));
+        };
+
+        rt::spawn(revocation::zmq_worker(
+            zmq_rx,
+            revocation_tx.clone(),
+            ip,
+            port,
+        ))
+        .map_err(Error::from)
+    } else {
+        rt::spawn(ok(())).map_err(Error::from)
+    };
+
+    let shutdown_task = rt::spawn(async move {
+        rt::signal::ctrl_c().await.unwrap(); //#[allow_ci]
+
+        info!("Shutting down keylime agent server");
+
+        // Shutdown tasks
+        let server_stop = server_handle.stop(true);
+        payload_tx.send(payloads::PayloadMessage::Shutdown);
+        keys_tx.send((keys_handler::KeyMessage::Shutdown, None));
+
+        #[cfg(feature = "with-zmq")]
+        zmq_tx.send(revocation::ZmqMessage::Shutdown);
+
+        revocation_tx.send(revocation::RevocationMessage::Shutdown);
+
```

Comment:
Do we also need to shutdown the connection to the TPM cleanly (e.g. clean all sessions and flush the keys) or is that handled by the TPM library?

Reply:
From the messages I get when using `RUST_LOG=trace`, I believe the TPM library takes care of the context when the variable is dropped:
```
 INFO  tss_esapi::context                 > Closing context.
 INFO  tss_esapi::context                 > Flushing handle 1078035600
 INFO  tss_esapi::context                 > Flushing handle 1078035599
 INFO  tss_esapi::context                 > Closing handle 1078035594
 INFO  tss_esapi::context                 > Context closed.
```